libmcount/mcount.c

```diff
@@ -1837,6 +1837,21 @@ static struct rb_root *deep_copy_triggers(struct rb_root *src)
 	return new;
 }
 
+/**
+ * swap_triggers - atomically swap the pointer to a filter rbtree and free the
+ * old one
+ * @old - pointer to the tree to deprecate
+ * @new - new version of the tree to use
+ */
+static void swap_triggers(struct rb_root **old, struct rb_root *new)
+{
+	struct rb_root *tmp;
+	tmp = __sync_val_compare_and_swap(old, *old, new);
+	sleep(1); /* RCU-like grace period */
```

Comment:
I'm not sure 1 second sleep is good enough.  But I don't have a good alternative now.  Maybe we can consider more sophisticated grace period handling later, but let's start with the simple approach.

Reply:
I was wondering if some kind of atomic read is required for the read side, but it seems `mcount_entry_filter_check()` only passes the root of the tree once.  Looks ok for now.